src/Symfony/Component/HttpFoundation/StreamedJsonResponse.php

```diff
@@ -0,0 +1,157 @@
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+/**
+ * StreamedJsonResponse represents a streamed HTTP response for JSON.
+ *
+ * A StreamedJsonResponse uses a structure and generics to create an
+ * efficient resource saving JSON response.
+ *
+ * It uses flush after a specified flush size to directly stream the data.
+ *
+ * @see flush()
+ *
+ * @author Alexander Schranz <alexander@sulu.io>
+ *
+ * Example usage:
+ *
+ * function loadArticles(): \Generator
+ *     // some streamed loading
+ *     yield ['title' => 'Article 1'];
+ *     yield ['title' => 'Article 2'];
+ *     yield ['title' => 'Article 3'];
+ * }),
+ *
+ * $response = new StreamedJsonResponse(
+ *     // json structure with generators in which will be streamed
+ *     [
+ *         '_embedded' => [
+ *             'articles' => loadArticles(), // any generator which you want to stream as list of data
+ *         ],
+ *     ],
+ * );
+ */
+class StreamedJsonResponse extends StreamedResponse
+{
+    public const DEFAULT_ENCODING_OPTIONS = JsonResponse::DEFAULT_ENCODING_OPTIONS;
+
+    private int $encodingOptions = self::DEFAULT_ENCODING_OPTIONS;
+
+    /**
+     * @param mixed[]                        $data      JSON Data containing PHP generators which will be streamed as list of data
+     * @param int                            $status    The HTTP status code (200 "OK" by default)
+     * @param array<string, string|string[]> $headers   An array of HTTP headers
+     * @param int                            $flushSize After every which item of a generator the flush function should be called
+     */
+    public function __construct(
+        private readonly array $data,
+        int $status = 200,
+        array $headers = [],
+        private int $flushSize = 100,
+    ) {
+        parent::__construct($this->stream(...), $status, $headers);
+
+        if (!$this->headers->get('Content-Type')) {
+            $this->headers->set('Content-Type', 'application/json');
+        }
+    }
+
+    private function stream(): void
+    {
+        $generators = [];
+        $structure = $this->data;
+
+        array_walk_recursive($structure, function (&$item, $key) use (&$generators) {
+            if ($item instanceof \Generator) {
+                // using uniqid to avoid conflict with eventually other data in the structure
+                $placeholder = uniqid('__placeholder_', true);
+                $generators[$placeholder] = $item;
+
+                $item = $placeholder;
+            }
+        });
+
+        $keys = array_keys($generators);
+
+        $jsonEncodingOptions = \JSON_THROW_ON_ERROR | $this->getEncodingOptions();
+        $structureText = json_encode($structure, $jsonEncodingOptions);
+
+        foreach ($keys as $key) {
+            [$start, $end] = explode('"'.$key.'"', $structureText, 2);
+
+            // send first and between parts of the structure
+            echo $start;
+
+            $count = 0;
+            echo '[';
```

Comment:
what if the generator returns a map and not a list ? Here, you are totally ignoring the keys

Reply:
As documented we only support `List` no maps are supported!

Reply:
:cry:

Reply:
The only way to detect if its a list would be to check if the first element is int `0`. Which will something like `["test", "key" => "value"]` also convert to a List. But could solve the problem with the `\Traversable` check.

Reply:
In case of mixed key type, the result type will be decided with the 1st item. The order of elements have a huge impact. I think it would be helpful to throw an exception in this situation. 

```php
function generateData(): \Generator {
    yield 0 => 'test';
    yield 'key' => 'value';
}
```

Output a list:
```json
[
    "test",
    "value",
]
```

When generator starts with a string key
```php
function generateData(): \Generator {
    yield 'key' => 'value';
    yield 0 => 'test';
}
```

Output a map:
```json
[
    "key" => "value",
    "0" => "test",
]
```

Reply:
We need unit tests for both cases.

Reply:
```php
function generateData(): \Generator {
    yield 'key' => 'value';
    yield 0 => 'test';
}
```

will result in `{"key": "value", 0: "test"}` which would be valid.

For the other case we maybe should do a `dev` only check to have no impact on it in production. Like it was done here: https://github.com/symfony/symfony/pull/40317

Reply:
Still we should keep in mind an exception doesn't add here a lot. Status Code 200 is already send and we are here already in the last line of the index.php which is not wrapped by the Kernel and so here no nice error message in any case will be shown.

Reply:
The current implementation is consistent with PHP `json_encode`, except when the 1st key is `0`, but the following keys are not sequential.

> will result in `{"key": "value", 0: "test"}` which would be valid.

Actually, this JSON cannot be decoded by [PHP](https://3v4l.org/EATha#v8.1.11) or Javascript `JSON.parse`. 
I propose to cast the key as string: https://github.com/alexander-schranz/symfony/pull/1/files?diff=unified&w=0#diff-88676ff0fb58128ade940542c46b371799cd3d2cb30f4b1d83d5e5beb165a661R115

> Still we should keep in mind an exception doesn't add here a lot. Status Code 200 is already send and we are here already in the last line of the index.php which is not wrapped by the Kernel and so here no nice error message in any case will be shown.

Silencing this error behavior is not a good option of DX.
May be a good use-case for a `trigger_error('...', E_USER_NOTICE)`, when the key type is inconsistent.

Reply:
With the usage of `JSON_NUMERIC_CHECK` this seems also be solved?

Reply:
We still have an inconsistency with `json_encode` as it is not possible to scan all keys for a map with the 1st key being `0`.

```php
$data = [0 => 'zero', 2 => 'two'];
json_encode($data); // {"0":"zero","2":"two"}
StreamJsonResponse($data); // ["zero","two"]
```